index.bs

```diff
@@ -1616,27 +1673,37 @@ partial interface MLGraphBuilder {
     **Returns:** an {{MLOperand}}. The permuted or transposed N-D tensor. 
 </div>
 
-## MLGraph ## {#api-mlgraph}
-The {{MLGraph}} interface represents a compiled computational graph. A compiled graph once constructed is immutable and cannot be subsequently changed.
+## MLTensor ## {#api-mltensor}
+The {{MLTensor}} interface represents a resource of a multidimensional array of numbers with a shape.
 
 <script type=idl>
-dictionary MLInput {
-  required (MLBufferView or WebGLTexture or GPUTexture) data;
-  sequence<long> dimensions;
-};
+typedef (MLBufferView or WebGLTexture or GPUTexture) MLResource;
 
-dictionary MLOutput {
-  (MLBufferView or WebGLTexture or GPUTexture) data;
-  sequence<long> dimensions;
+[SecureContext, Exposed=Window]
+interface MLTensor {
+  // Construct a tensor from a resource
+  constructor(MLResource resource, optional sequence<long> dimensions);
+
+  // The tensor dimensions
+  sequence<long> dimensions();
+
+  // The underlying resource of the tensor
+  MLResource resource();
+
+  // Asynchronously download the result of a computation onto a typed array.
+  Promise<ArrayBufferView> data();
 };
+</script>
 
-typedef record<DOMString, MLInput> MLNamedInputs;
-typedef record<DOMString, MLOutput> MLNamedOutputs;
+## MLGraph ## {#api-mlgraph}
+The {{MLGraph}} interface represents a compiled computational graph. A compiled graph once constructed is immutable and cannot be subsequently changed.
+
+<script type=idl>
+typedef record<DOMString, MLTensor> MLNamedTensors;
 
 [SecureContext, Exposed=Window]
 interface MLGraph {
-  Promise<MLNamedOutputs> compute(MLNamedInputs inputs, 
-                                  optional MLNamedOutputs outputs = {});
+  MLNamedTensors compute(MLNamedTensors inputs, optional MLNamedTensors outputs = {});
```

Comment:
The `compute` method probably could be overrode for different device types and different output allocations (newly allocated or pre-allocated), e.g.,
```js
typedef record<DOMString, MLTensor> MLNamedTensors;
typedef record<DOMString, MLGPUTensor> MLNamedGPUTensors;

interface MLGraph {
  // MLContext with any device types, allocate output `MLTensor`s for listed output names.
  MLNamedTensors compute(MLNamedTensors inputs, optional sequence<DOMString> outputNames);

  // MLContext with any device types, use pre-allocated output 'MLTensor`s.
  void compute(MLNamedTensors inputs, MLNamedTensors outputs);

  // MLContext with GPUDevice or WebGLRenderingContext device type, allocate output `MLGPUTensor`s for listed output names.
  MLNamedGPUTensors compute(MLNamedGPUTensors inputs, optional sequence<DOMString> outputNames);

  // MLContext with GPUDevice or WebGLRenderingContext device type, use pre-allocated output 'MLTensor`s.
  void compute(MLNamedGPUTensors inputs, MLNamedGPUTensors outputs);
};
```

Reply:
For the CPU - WebAssembly scenario of [Operation-specific API](https://github.com/webmachinelearning/proposals/issues/2#issuecomment-773441501), the framework could create `MLTensor`s with pre-allocated Wasm memory, e.g.,

```js
const inputs = {'x': new MLTensor(input_tensor_info.f32())};
const outputs = {'y': new MLTensor(output_tensor_info.f32_write())};
conv2d.compute(inputs, outputs);
await outputs.y.data()
// The results have been downloaded into the memory pointed by output_tensor_info.f32_write().
```

For GPU scenario, the framework could create `MLGPUTensor`s with WebGLTextures, e.g.,
```js
const inputs = {'x': new MLGPUTensor(inputTextureData.texture)};
const outputs = {'y': new MLGPUTensor(outputTextureData.texture)};
conv2d.compute(inputs, outputs);
// Post-processing with outputTextureData.texture
```

@wchao1115 @pyu10055 , does it look good?

Reply:
@huningxin In WebGL, there is no way to share textures across contexts, is the input textures limited to a GPU context created during compilation or users can specify a context during graph creation?

Reply:
@pyu10055 the assumption is that the app must create an MLContext with the WebGL context that creates the texture, which later passed back as an input to the graph created by the MLContext. It is generally known that a resource should not be shared across different GPU contexts. In DirectX, you can achieve cross-adapter sharing of resources, but you need to explicitly create a shared resource handle and manage the lifetime of the shared resource explicitly. But at the Web API level, I agree that it is best to make this simpler and simply not allow cross-adapter sharing.

Reply:
@pyu10055 

> users can specify a context during graph creation?

yes, user can create a `MLContext` by specifying a `WebGLRenderingContext`, as [ML.createContext](https://webmachinelearning.github.io/webnn/#api-ml)
```js
MLContext createContext(WebGLRenderingContext glContext);
```

Reply:
@huningxin I suggest only to have 1 method that requires the caller to pre-allocate the output buffer. It's much simpler that way i.e.

```javascript
typedef record<DOMString, MLTensor> MLNamedTensors;

interface MLGraph {
  // MLContext with any device types, use pre-allocated output 'MLTensor`s.
  void compute(MLNamedTensors inputs, MLNamedTensors outputs);
};
```

I wonder how practical the optional output case (Example 5) really is. For a multi-output topology, not requiring all of the outputs to be produced at execution is uncommon. The optimization gain from not having to specify an optional output is going to be minimal relative to the cost of executing the entire topology anyway. We should probably remove this use case and simplify the API.

Also, there is no need to special-case the GPU resources tensor as the `MLTensor' already wraps those types nicely. See the 'Device Selection` section for more detail as to when different type of resource is allowed.

Reply:
> I wonder how practical the optional output case (Example 5) really is.

The use case came from https://github.com/webmachinelearning/webnn/issues/105 @pyu10055 . With the new `MLTensor` interface, user code can only download the data of a specific output. This would help avoid the overhead of downloading other outputs that user code doesn't need.

Reply:
Sure. And this would still work, right? 

```Javascript
const d = builder.matmul(a, b);
const e = builder.add(d, c);
const graph = await builder.build({d, e});
const output = {'d': new MLTensor(new Float32Array(sizeOfShape([3, 3])))};
graph.compute(inputs, output);
let data = await output.d.data();
```

Reply:
right.